src/lib/notifier.c

```diff
@@ -14,82 +14,164 @@
+		for (i = 0; i < NOTIFIER_ID_ALL; i++)
+			notifier_unregister(receiver, caller, i);
+		return;
+	}
 
 	spin_lock(notify->lock);
-	list_item_del(&notifier->list);
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
+		handle = container_of(wlist, struct callback_handle, list);
+		if ((!receiver || handle->receiver == receiver) &&
+		    (!caller || handle->caller == caller)
+		   ) {
+			list_item_del(&handle->list);
+			rfree(handle);
+		}
+	}
 	spin_unlock(notify->lock);
 }
 
-void notifier_notify(void)
+static void notifier_notify(void *caller, enum notify_id type, void *data)
 {
 	struct notify *notify = *arch_notify_get();
 	struct list_item *wlist;
-	struct notifier *n;
-
-	if (!list_is_empty(&notify->list)) {
-		dcache_invalidate_region(&_notify_data, sizeof(_notify_data));
-		dcache_invalidate_region(_notify_data.data,
-					 _notify_data.data_size);
-
-		/* iterate through notifiers and send event to
-		 * interested clients
-		 */
-		list_for_item(wlist, &notify->list) {
-			n = container_of(wlist, struct notifier, list);
-			if (n->id == _notify_data.id)
-				n->cb(_notify_data.message, n->cb_data,
-				      _notify_data.data);
-		}
+	struct list_item *tlist;
+	struct callback_handle *handle;
+
+	/* iterate through notifiers and send event to
+	 * interested clients
+	 */
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
+		handle = container_of(wlist, struct callback_handle, list);
+		if (!caller || !handle->caller || handle->caller == caller)
+			handle->cb(handle->receiver, type, data);
 	}
 }
 
-void notifier_event(struct notify_data *notify_data)
+void notifier_notify_remote(void)
 {
 	struct notify *notify = *arch_notify_get();
+	struct notify_data *notify_data = &_notify_data[cpu_get_id()];
+
+	dcache_invalidate_region(notify_data, sizeof(*notify_data));
+	if (!list_is_empty(&notify->list[notify_data->type])) {
+		dcache_invalidate_region(notify_data->data,
+					 notify_data->data_size);
+		notifier_notify(notify_data->caller, notify_data->type,
+				notify_data->data);
+	}
+}
+
+void notifier_event(void *caller, enum notify_id type, uint32_t core_mask,
+		    void *data, uint32_t data_size)
+{
+	struct notify_data *notify_data;
 	struct idc_msg notify_msg = { IDC_MSG_NOTIFY, IDC_MSG_NOTIFY_EXT };
 	int i;
 
-	spin_lock(notify->lock);
-
-	_notify_data = *notify_data;
-	dcache_writeback_region(_notify_data.data, _notify_data.data_size);
-	dcache_writeback_region(&_notify_data, sizeof(_notify_data));
-
 	/* notify selected targets */
 	for (i = 0; i < PLATFORM_CORE_COUNT; i++) {
-		if (_notify_data.target_core_mask & (1 << i)) {
+		if (core_mask & NOTIFIER_TARGET_CORE_MASK(i)) {
 			if (i == cpu_get_id()) {
-				notifier_notify();
+				notifier_notify(caller, type, data);
 			} else if (cpu_is_core_enabled(i)) {
 				notify_msg.core = i;
+				notify_data = &_notify_data[i];
+				notify_data->caller = caller;
+				notify_data->type = type;
+				notify_data->data = data;
```

Comment:
I've just realized that some of the data here is passed from the local stack, so with asynchronous messages to other cores it still won't work correctly. Although it's not an issue right now and this problem has existed before your changes, we need to remember to have it fixed in the near future (we could allocate array with fixed max size and memcpy the data). For the sake of this PR such changes are not required.

Reply:
@tlauda Notification payload evaluated on the same core might be located on stack, without causing any issues, and for multicore notifications, I've made sure that event payload is allocated within the 'caller', or is the caller itself.
So while there is the possibility to misuse multicore events with payload on stack, such instance doesn't exist at the moment. We might possibly add assert to verify that payload is outside of stack memory, to prohibit such future usages.

Reply:
IDK if it's ok with kernel guidelines, but maybe we should add comments like `// FXIME` or `// TODO` in cases like this, so when someone revisits that part of code he will see potential problem instantly.

Reply:
Some sort of `/* TODO: Check if attempting to pass on-stack data to a different core */ somewhere here. Although when talking about cross-core calls like with this notifier I'd expect callers to _be_ aware of such issues.